### PR TITLE
Margin ranking loss backward

### LINKED ISSUE #2348: # Bug: margin_ranking_loss 缺少 Autograd 注册导致梯度链断裂
## 问题概述

**严重程度:** 高 (P1)
**问题类型:** FlagGems算子实现缺陷
**影响范围:** 所有依赖 margin_ranking_loss backward 的训练场景

## 问题描述

FlagGems 的 `margin_ranking_loss` 算子只注册了前向计算，没有注册对应的 autograd kernel。当 backward 传播经过此算子时，梯度链断裂，导致输入 tensor 的梯度为 None。

PyTorch 在 backward 时会发出以下警告：
```
aten::margin_ranking_loss: an autograd kernel was not registered to the Autograd key(s) 
but we are trying to backprop through it. This may lead to silently incorrect behavior.
This behavior is deprecated and will be removed in a future version of PyTorch.
```

## 复现步骤

```python
import torch
import flag_gems

flag_gems.enable()

anchor = torch.randn(4, device="cuda", requires_grad=True)
positive = torch.randn(4, device="cuda", requires_grad=True)
label = torch.ones(4, device="cuda")

loss_fn = torch.nn.MarginRankingLoss(margin=1.0)
loss = loss_fn(anchor, positive, label)
loss.backward()

print(anchor.grad)   # None — 预期应有梯度值
print(positive.grad)  # None — 预期应有梯度值
```

## 根因分析

**算子文件位置:** `src/flag_gems/ops/margin_ranking_loss.py`

**注册方式:** 在 `src/flag_gems/__init__.py:254` 通过 `("margin_ranking_loss", margin_ranking_loss)` 注册到 `aten::margin_ranking_loss`。

当前实现只提供了前向 kernel (`_margin_ranking_loss_kernel`)，通过 Triton 实现。但 `aten::margin_ranking_loss` 在 PyTorch 中是一个 composite 算子（本身由更基本的操作组合而成），原生 PyTorch 是通过 `CompositeImplicitAutograd` dispatch key 来自动获得 backward 支持的。

FlagGems 覆盖了这个算子但没有注册到 `Autograd` key，导致 autograd 路径被阻断。

## 是否需要 Fix

**需要** — 这是一个在实际训练场景中必然触发的缺陷：
- margin_ranking_loss 广泛用于排序学习、对比学习、度量学习等场景
- 这些场景必然需要 backward，否则无法训练
- 当前实现会**静默**地丢失梯度（只有 warning，不会报错），可能导致训练看似正常但模型实际不学习

## 建议修复方案

方案 1（推荐）：将 `margin_ranking_loss` 实现为 composite 算子，分解为 FlagGems 已经支持的基本操作 (sub, mul, clamp, mean/sum)，让 autograd 自动处理。

方案 2：实现自定义的 `torch.autograd.Function`，手动编写 forward 和 backward 的 Triton kernel。

## 单测覆盖分析

**现有单测:** `tests/test_special_ops.py:2466` — `test_accuracy_margin_ranking_loss`

**为什么没有发现该问题:**
- 单测只验证了前向精度（`gems_assert_close(res_out, ref_out, dtype)`），没有测试 backward
- 没有验证 `requires_grad=True` 输入通过算子后梯度是否正确传播
- 单测的设计模式是"前向精度对比"，不包含梯度正确性验证

**组合测例为什么能发现:**
- 组合测例将 `margin_ranking_loss` 放入多任务损失模型中
- 模型训练需要 `loss.backward()`，自然触发了 backward 路径
- 验证了所有 `requires_grad=True` 输入的 `.grad` 属性
- 这反映了组合测例相比单元测试更贴近实际使用场景的优势

## 影响的使用场景

| 场景 | 影响程度 | 说明 |
|------|---------|------|
| 排序学习 (Learning to Rank) | 高 | margin_ranking_loss 是核心损失函数 |
| 对比学习 | 中 | 常用于构建 triplet loss 等变体 |
| 推荐系统 | 中 | pairwise ranking 场景 |
| 度量学习 | 中 | margin-based loss 的基础 |

### PR DESCRIPTION
**Pull request by YY4994**
### PR Category
Operator | OP Test

### Type of Change
Bug Fix

### Description
Rewrite `margin_ranking_loss` as a composite of differentiable primitive ops so autograd can trace through the computation graph naturally. Add backward tests covering all reduction modes, shapes, and float dtypes.

### Issue

- Resolves #2348

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### CODE REVIEW

**Review by tengqm — COMMENTED** [review]
Confused by this change.
Are we reworking the existing `margin_ranking_loss` function or adding a new function that handles the backward situation?
What is the intent of this PR?

**Comment by YY4994** [comment]
> Confused by this change. Are we reworking the existing `margin_ranking_loss` function or adding a new function that handles the backward situation? What is the intent of this PR?

This PR fixes the issue that margin_ranking_loss cannot perform backward propagation. For bug details, please see BUG#2348.

**Comment by tengqm** [comment]
> > Confused by this change. Are we reworking the existing `margin_ranking_loss` function or adding a new function that handles the backward situation? What is the intent of this PR?
> 
> This PR fixes the issue that margin_ranking_loss cannot perform backward propagation. For bug details, please see BUG#2348.

Thanks for the quick response. So you mean the existing code for `margin_ranking_loss` has problems with backward case, and this PR **reworks** the operator so that it can handle both forward and backward case. Am I understanding this correctly?

**Comment by YY4994** [comment]
> > > Confused by this change. Are we reworking the existing `margin_ranking_loss` function or adding a new function that handles the backward situation? What is the intent of this PR?
> > 
> > 
> > This PR fixes the issue that margin_ranking_loss cannot perform backward propagation. For bug details, please see BUG#2348.
> 
> Thanks for the quick response. So you mean the existing code for `margin_ranking_loss` has problems with backward case, and this PR **reworks** the operator so that it can handle both forward and backward case. Am I understanding this correctly?

yes

**Review by tengqm — APPROVED** [review]

**Comment by tengqm** [comment]
@YY4994 Why is this closed?

**Comment by YY4994** [comment]
> @YY4994 Why is this closed?

I found that this modification did not completely eliminate the UserWarning. I tried a new approach, but it turned out that just modifying the operator could not eliminate the warning.
